File: ios/ScheduleWidget/ScheduleWidgetData.swift
import Foundation

let widgetGroupId = "group.com.hmncorp.familyplanner"

struct ScheduleItem: Decodable {
    let id: String
    let title: String
    let scheduledAt: String?
    let allDay: Bool
    let colorHex: String
    let isCompleted: Bool
    let groupId: String? // nil = 개인 일정

    var timeText: String {
        if allDay { return "종일" }
        guard let scheduledAt = scheduledAt, let tIndex = scheduledAt.firstIndex(of: "T") else {
            return "-"
        }
        let timePart = scheduledAt[scheduledAt.index(after: tIndex)...]
        return String(timePart.prefix(5))
    }
}

/// 월간 데이터의 날짜별 항목. 일정 상세(item)를 함께 들고 있어
/// 위젯이 특정 날짜를 탭했을 때 서버 호출 없이 그 날짜의 일정 목록을 그릴 수 있다.
struct MonthDateEntry: Decodable {
    let date: String
    let item: ScheduleItem

    var groupId: String? { item.groupId }

    private enum CodingKeys: String, CodingKey {
        case date
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        date = try container.decode(String.self, forKey: .date)
        item = try ScheduleItem(from: decoder)
    }
}

struct MonthScheduleData: Decodable {
    let year: Int
    let month: Int
    let entries: [MonthDateEntry]
}

struct WidgetGroup: Decodable, Identifiable, Hashable {
    let id: String
    let name: String
    let colorHex: String
}

enum ScheduleWidgetData {
    private static let todayTasksKey = "today_tasks"
    private static let monthTasksKey = "month_tasks"
    private static let groupsKey = "widget_groups"

    static func readTodayTasks() -> [ScheduleItem] {
        guard let defaults = UserDefaults(suiteName: widgetGroupId),
            let raw = defaults.string(forKey: todayTasksKey),
            let data = raw.data(using: .utf8)
        else { return [] }

        return (try? JSONDecoder().decode([ScheduleItem].self, from: data)) ?? []
    }

    /// 동기화된 지난달/이번달/다음달(-1~+1) 목록을 모두 읽는다
    static func readMonths() -> [MonthScheduleData] {
        guard let defaults = UserDefaults(suiteName: widgetGroupId),
            let raw = defaults.string(forKey: monthTasksKey),
            let data = raw.data(using: .utf8)
        else { return [] }

        struct Root: Decodable { let months: [MonthScheduleData] }
        return (try? JSONDecoder().decode(Root.self, from: data))?.months ?? []
    }

    /// [readMonths] 결과에서 특정 연/월을 찾는다. 없으면 빈 entries를 가진 데이터를 반환한다.
    static func readMonth(year: Int, month: Int) -> MonthScheduleData {
        readMonths().first { $0.year == year && $0.month == month }
            ?? MonthScheduleData(year: year, month: month, entries: [])
    }

    static func readGroups() -> [WidgetGroup] {
        guard let defaults = UserDefaults(suiteName: widgetGroupId),
            let raw = defaults.string(forKey: groupsKey),
            let data = raw.data(using: .utf8)
        else { return [] }

        return (try? JSONDecoder().decode([WidgetGroup].self, from: data)) ?? []
    }
}

/// 위젯 인스턴스별 그룹 필터. selectedGroupIds가 nil이면 전체 포함(필터 없음).
struct WidgetScheduleFilter {
    let selectedGroupIds: Set<String>?
    let includePersonal: Bool

    static let all = WidgetScheduleFilter(selectedGroupIds: nil, includePersonal: true)

    func matches(groupId: String?) -> Bool {
        guard let groupId = groupId else { return includePersonal }
        guard let selectedGroupIds = selectedGroupIds else { return true }
        return selectedGroupIds.contains(groupId)
    }
}

extension Array where Element == ScheduleItem {
    func applyFilter(_ filter: WidgetScheduleFilter) -> [ScheduleItem] {
        self.filter { filter.matches($0.groupId) }
    }
}

extension MonthScheduleData {
    func markedDates(_ filter: WidgetScheduleFilter) -> Set<String> {
        Set(entries.filter { filter.matches($0.groupId) }.map { $0.date })
    }

    /// 특정 날짜(dateKey, "yyyy-MM-dd")의 일정 목록을 시간순으로 반환한다
    func items(forDate dateKey: String, filter: WidgetScheduleFilter) -> [ScheduleItem] {
        entries
            .filter { $0.date == dateKey && filter.matches($0.groupId) }
            .map { $0.item }
            .sorted { ($0.scheduledAt ?? "") < ($1.scheduledAt ?? "") }
    }
}

File: android/app/src/main/kotlin/com/hmncorp/familyplanner/widget/ScheduleWidgetData.kt
package com.hmncorp.familyplanner.widget

import android.content.Context
import es.antonborri.home_widget.HomeWidgetPlugin
import org.json.JSONArray
import org.json.JSONObject

data class ScheduleItem(
    val id: String,
    val title: String,
    val scheduledAt: String?,
    val allDay: Boolean,
    val colorHex: String,
    val isCompleted: Boolean,
    val groupId: String?, // null = 개인 일정
)

data class MonthDateEntry(
    val date: String,
    val item: ScheduleItem,
) {
    val groupId: String? get() = item.groupId
}

data class MonthScheduleData(
    val year: Int,
    val month: Int,
    val entries: List<MonthDateEntry>,
)

data class WidgetGroup(
    val id: String,
    val name: String,
    val colorHex: String,
)

/** Flutter의 HomeWidgetService가 저장한 JSON을 읽어 위젯이 쓰기 좋은 형태로 파싱 */
object ScheduleWidgetData {
    private const val KEY_TODAY_TASKS = "today_tasks"
    private const val KEY_MONTH_TASKS = "month_tasks"
    private const val KEY_GROUPS = "widget_groups"

    fun readTodayTasks(context: Context): List<ScheduleItem> {
        val raw = HomeWidgetPlugin.getData(context).getString(KEY_TODAY_TASKS, null) ?: return emptyList()
        return try {
            val array = JSONArray(raw)
            (0 until array.length()).map { i ->
                val obj = array.getJSONObject(i)
                ScheduleItem(
                    id = obj.optString("id", ""),
                    title = obj.optString("title", ""),
                    scheduledAt = obj.optString("scheduledAt", null),
                    allDay = obj.optBoolean("allDay", false),
                    colorHex = obj.optString("colorHex", "#2196F3"),
                    isCompleted = obj.optBoolean("isCompleted", false),
                    groupId = if (obj.isNull("groupId")) null else obj.optString("groupId", null),
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    /** 동기화된 지난달/이번달/다음달(-1~+1) 목록을 모두 읽는다 */
    fun readMonths(context: Context): List<MonthScheduleData> {
        val raw = HomeWidgetPlugin.getData(context).getString(KEY_MONTH_TASKS, null) ?: return emptyList()
        return try {
            val root = JSONObject(raw)
            val monthsArray = root.optJSONArray("months") ?: return emptyList()
            (0 until monthsArray.length()).mapNotNull { i ->
                val obj = monthsArray.getJSONObject(i)
                val year = obj.optInt("year", -1)
                val month = obj.optInt("month", -1)
                if (year < 0 || month < 0) return@mapNotNull null
                val rawEntries = obj.optJSONArray("entries") ?: JSONArray()
                val entries = (0 until rawEntries.length()).map { j ->
                    val entryObj = rawEntries.getJSONObject(j)
                    MonthDateEntry(
                        date = entryObj.optString("date", ""),
                        item = ScheduleItem(
                            id = entryObj.optString("id", ""),
                            title = entryObj.optString("title", ""),
                            scheduledAt = entryObj.optString("scheduledAt", null),
                            allDay = entryObj.optBoolean("allDay", false),
                            colorHex = entryObj.optString("colorHex", "#2196F3"),
                            isCompleted = entryObj.optBoolean("isCompleted", false),
                            groupId = if (entryObj.isNull("groupId")) null else entryObj.optString("groupId", null),
                        ),
                    )
                }
                MonthScheduleData(year, month, entries)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    /** [readMonths] 결과에서 특정 연/월을 찾는다. 없으면 빈 entries를 가진 데이터를 반환한다. */
    fun readMonth(context: Context, year: Int, month: Int): MonthScheduleData {
        return readMonths(context).firstOrNull { it.year == year && it.month == month }
            ?: MonthScheduleData(year, month, emptyList())
    }

    fun readGroups(context: Context): List<WidgetGroup> {
        val raw = HomeWidgetPlugin.getData(context).getString(KEY_GROUPS, null) ?: return emptyList()
        return try {
            val array = JSONArray(raw)
            (0 until array.length()).map { i ->
                val obj = array.getJSONObject(i)
                WidgetGroup(
                    id = obj.optString("id", ""),
                    name = obj.optString("name", ""),
                    colorHex = obj.optString("colorHex", "#2196F3"),
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }
}

/**
 * 위젯 인스턴스별 그룹 필터 설정.
 * [selectedGroupIds] null = 전체 그룹 포함(필터 없음), 빈 set = 그룹 전부 해제(개인만),
 * 값 있음 = 해당 그룹만 포함.
 * [includePersonal] 개인 일정 포함 여부
 */
data class WidgetFilter(
    val selectedGroupIds: Set<String>?,
    val includePersonal: Boolean,
) {
    fun matches(groupId: String?): Boolean {
        if (groupId == null) return includePersonal
        if (selectedGroupIds == null) return true
        return selectedGroupIds.contains(groupId)
    }

    companion object {
        val ALL = WidgetFilter(selectedGroupIds = null, includePersonal = true)
    }
}

fun List<ScheduleItem>.applyFilter(filter: WidgetFilter): List<ScheduleItem> =
    filter { filter.matches(it.groupId) }

fun MonthScheduleData.markedDates(filter: WidgetFilter): Set<String> =
    entries.filter { filter.matches(it.groupId) }.map { it.date }.toSet()

/** 특정 날짜(dateKey, "yyyy-MM-dd")의 일정 목록을 시간순으로 반환한다 */
fun MonthScheduleData.itemsForDate(dateKey: String, filter: WidgetFilter): List<ScheduleItem> =
    entries
        .filter { it.date == dateKey && filter.matches(it.groupId) }
        .map { it.item }
        .sortedWith(compareBy(nullsLast()) { it.scheduledAt })
